pysynchronizer/cmd.py

```diff
@@ -0,0 +1,207 @@
+    def __init__(self):
+        Cmd.__init__(self)
+
+    def run(self, cmd_str=""):
+        if cmd_str == "":
+            self.cmdloop()
+        else:
+            self.onecmd(cmd_str)
+
+    def do_exit(self, args):
+        return -1
+
+class SyncCmd(BaseCmd):
+    def __init__(self, cmd_str=""):
+        BaseCmd.__init__(self)
+        self.prompt = "sync> "
+
+    def do_xenmgr(self, arg_str):
+        try:
+            XenMgrCmd().run(arg_str)
+        except CommuincationError as err:
+            print("Connection failed:\n\t%s\n" % err)
+        except Exception as err:
+            print("Unexpected exception:\n\t%s\n" % err)
+
+    def do_vm(self, arg_str):
+        args = arg_str.split()
+
+        try:
+            if args:
+                vc = VmCmd(args[0])
+                arg_str = " ".join(args[1:])
+            else:
+                vc = VmCmd()
+                arg_str = ""
+
+            vc.run(arg_str)
+        except CommuincationError as err:
+            print("Connection failed:\n\t%s\n" % err)
+        except Exception as err:
+            print("Unexpected exception:\n\t%s\n" % err)
+
+class XenMgrCmd(BaseCmd):
+    def __init__(self):
+        BaseCmd.__init__(self)
+
+        self.prompt = "xemmgr> "
+        self.xenmgr = XenMgr()
+
+    def do_list_vm(self, arg_str):
+        rows = [ [
+            "Name",
+            "UUID",
+        ] ]
+
+        vm_list = self.xenmgr.vm_by_name()
+        for name in vm_list:
+            rows.append([name, vm_list[name]])
+
+        column_print(rows)
+
+    def do_create_vm(self ,arg_str):
+        pass
+
+    def do_delete_vm(self, arg_str):
+        ref = arg_str.split()[0]
+        vm_uuid = self.xenmgr.find_vm(ref)
+
+        if vm_uuid:
+            vm = VM(vm_uuid)
+            vm.delete()
+        else:
+            print("unable to find vm: %s" % ref)
+
+    def do_upgrade(self, args):
+        if not self.xenmgr.upgrade(args.split()[0]):
+            print("upgrade failed")
+
+class VmCmd(BaseCmd):
+    def __init__(self, arg_str):
+        BaseCmd.__init__(self)
+
+        self.prompt = "vm> "
+
+        if len(arg_str) > 0:
+            self.select_vm(arg_str.split()[0])
+
+    def select_vm(self, vm_ref):
+        vm_uuid = XenMgr().find_vm(vm_ref)
+        if vm_uuid:
+            self.vm = VM(vm_uuid)
+            self.name = self.vm.get_property("name")
+            self.prompt = "vm[%s]> " % self.name
+        else:
+            self.vm = None
+            self.prompt = "vm> "
+
+    def do_select(self, arg_str):
+        self.vm = None
+        self.prompt = "vm> "
+
+        args = arg_str.split()
+
+        self.select_vm(args[0])
+        if self.vm is None:
+            print("Failed to select VM %s" % args[0])
+
+    def do_domstore(self, arg_str):
+        args = arg_str.split()
+        if not args:
+            print("usage: domstore get|set key [value]")
+            return
+
+        cmd, args = args[0], args[1:]
+
+        if cmd == "get":
```

Comment:
This is because of how the Cmd class works. It's main use is to provide an interactive command interpreter and so the module is that It passes the remainder of the command string as the second parameter to your method. The argparse library is explicitly for parsing all the command line arguments passed to python/python script.

As a result the sync-cmd can be ran in interactive mode or can be passed a command string at the command line to do a single operation.